### PR TITLE
Canonical url Fix: canonical url should not change. Now uses filegrouper attributes 

### PR DESCRIPTION
**Pull request by NarenderRajuB**
Canonical url Fix: canonical url now uses filegrouper attributes.

### CODE REVIEW

File: djangocms_versioning_filer/monkeypatch/admin/fileadmin.py [hunk 1]
```diff
@@ -5,23 +5,6 @@
 from ...helpers import check_file_label_exists_in_folder
 
 
-def save_model(func):
```
Comment (Aiky30) [inline djangocms_versioning_filer/monkeypatch/admin/fileadmin.py:8]: You can't delete this, it's not part of this PR!

File: djangocms_versioning_filer/monkeypatch/models.py [hunk 2]
```diff
@@ -30,6 +48,31 @@ def is_file_content_valid_for_revert(version, user):
         )
 
 
+def model_save(func):
+    def inner(self, *args, **kwargs):
+        func(self, *args, **kwargs)
+        if self.__class__ == File:
```
Comment (Aiky30) [inline djangocms_versioning_filer/monkeypatch/models.py:54]: I don't understand how this is supposed to work. This code won't be ran because the original function will have completed by this line.

File: djangocms_versioning_filer/monkeypatch/models.py [hunk 3]
```diff
@@ -30,6 +48,31 @@ def is_file_content_valid_for_revert(version, user):
...
+            # what should we do now?
+            # maybe this has a subclass, but is being saved as a File instance
+            # anyway. do we need to go check all possible subclasses?
+            pass
+        elif issubclass(self.__class__, File):
+            self._file_type_plugin_name = self.__class__.__name__
+        if self._old_is_public != self.is_public and self.pk:
+            self._move_file()
+            self._old_is_public = self.is_public
+        if self.id and self.grouper and not self.grouper.canonical_file_id:
```
Comment (Aiky30) [inline djangocms_versioning_filer/monkeypatch/models.py:64]: I'm pretty sure that this should still work in the Admin save_model method because the Versioning query manager override is on the Admin get_queryset method. Why is this logic here?

**Review by Aiky30 — CHANGES_REQUESTED** [review]

File: djangocms_versioning_filer/monkeypatch/models.py (hunk 3 above)
Comment (NarenderRajuB) [reply djangocms_versioning_filer/monkeypatch/models.py:64]: if we are using file_Admin monkey patch in tests when use this 
file_obj = self.create_file_obj(
                original_filename='test-test.doc',
                folder=self.file.folder,
                grouper=grouper,
                publish=False,
            )
file_obj.grouper.canonical_file_id is empty

Comment (Aiky30) [reply djangocms_versioning_filer/monkeypatch/models.py:64]: Ok so why is it empty when using the admin monkey patch vs the model save monkey patch?

It is populated in the model save monkey patch because you are running after the save method has been executed because your monkey patch runs the super() before which means an id does exist. This is why one is working and the other doesn't, you're not doing the same thing so it's not the same. You need to execute the save() then try and add the grouper in the admin monkey patch to get the same result!

Comment (Aiky30) [reply djangocms_versioning_filer/monkeypatch/models.py:64]: This monkey patch (model.py) executes the code twice! It runs twice because you execute func() before your code which means that all of the code in models.py save() runs first then your code runs. That's what func() is, its the raw code you're monkey patching.

File: djangocms_versioning_filer/monkeypatch/admin/fileadmin.py (hunk 1 above)
Comment (NarenderRajuB) [reply djangocms_versioning_filer/monkeypatch/admin/fileadmin.py]: revertedthis change back

File: djangocms_versioning_filer/monkeypatch/models.py (hunk 3 above)
Comment (NarenderRajuB) [reply djangocms_versioning_filer/monkeypatch/models.py]: I have reverted back the cahnge to fileadmin now. But still when a draft version is published , it should create published version where the grouper canonical_file_id is updated. But even when we can file_obj.save() the monkey patch is not called.

File: djangocms_versioning_filer/monkeypatch/admin/fileadmin.py [hunk 4]
```diff
@@ -9,7 +10,11 @@ def save_model(func):
     """Override the FileAdmin save_model method"""
     def inner(self, request, obj, form, change):
         func(self, request, obj, form, change)
-
+        if isinstance(obj, File):
```
Comment (Aiky30) [inline djangocms_versioning_filer/monkeypatch/admin/fileadmin.py:13]: This is better.

@NarenderRajuB I don't think that we need to check if this is a File object here because we do a check below if the object has an id and a grouper attached. If the model was say Grouper then it wouldn't have obj.grouper.

If you added this because an error occurred we may have to run a check that the object has the attribute grouper. Maybe: 

grouper = getattr(obj, "grouper", False)

Comment (Aiky30) [reply djangocms_versioning_filer/monkeypatch/admin/fileadmin.py]: Due to my misunderstanding of when save_model is executed please add the following logic to the file monkeypatch/models.py.

```
def save(func):
    """Override the FileAdmin save_model method"""

    def inner(self, *args, **kwargs):
        func(self, *args, **kwargs)

        grouper = getattr(self, "grouper", False)

        if self.id and grouper and not grouper.canonical_file_id:
            grouper.canonical_file_id = self.id
            grouper.save()

        return 
    return inner


filer.models.File.save = save(  # noqa: E305
    filer.models.File.save
)
```

Comment (NarenderRajuB) [reply djangocms_versioning_filer/monkeypatch/admin/fileadmin.py]: Updated

**Review by Aiky30 — APPROVED** [review]
